csharp/src/Apache.Arrow/Field.cs

```diff
@@ -45,5 +45,41 @@ public partial class Field
             IsNullable = nullable;
             Metadata = metadata?.ToDictionary(kv => kv.Key, kv => kv.Value);
         }
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as Field);
+        }
+
+        public bool Equals(Field other)
+        {
+            if (other != null && this.Name == other.Name && this.IsNullable == other.IsNullable &&
+                this.DataType.TypeId == other.DataType.TypeId && HasMetadata == other.HasMetadata)
+            {
+                if (this.HasMetadata && other.HasMetadata)
+                {
+                    return this.Metadata.Keys.Count() == other.Metadata.Keys.Count() &&
+                           this.Metadata.Keys.All(k => other.Metadata.ContainsKey(k) && this.Metadata[k] == other.Metadata[k]) &&
+                           other.Metadata.Keys.All(k => this.Metadata.ContainsKey(k) && other.Metadata[k] == this.Metadata[k]);
```

Comment:
This makes Equals quite expensive.

Does Field actually need to be `IEquatable<Field>`?  What benefit does it provide?  I'm curious because this PR is titled being about tests but is updating public API in the src.

If for some reason it is required, it'd be better to avoid LINQ as part of this, calls to Count(), etc.  There are faster ways to achieve this that allocate less.  In particular, `Metadata` is being exposed as an `IReadOnlyDictionary<>`, but we create it as a `Dictionary<>` and so could operate on it as such; that would mean changing the property from `public IReadOnlyDictionary<string, string> Metadata { get; }` to `public IReadOnlyDictionary<string, string> Metadata => _metadata;` and declaring a field `private readonly Dictionary<string, string> _metdata;` that's then initialized in the ctor and that we directly acces in places like this.

Reply:
Yup, I agree that this makes Equals expensive. On the other hand, I don't see any faster paths to compare 2 dictionaries when I need to ensure that the data in the 2 dictionaries are exactly the same though. 

I could add a parameter to the method to skip the metadata checks by default. That would make the method faster, but that means 2 Field objects, one with metadata and one without, would evaluate to equal => GetHashCode() would need similar behavior. I'm ok with doing that if that seems acceptable? Also, I haven't seen real big datasets yet, so this may not be true, but I think the metadata dictionary will be pretty small(if even present) for most cases?

For why Field needs IEquatable<>, TypeTests::Equality() calls 
`Assert.True(f0_nullable.Equals(f0_other));`
Granted, this is a trivial test case now, but I need this to build tests around other things. For ex: I'm writing a test now to roundtrip a Schema to a stream to test I/O. After the read back from the stream, I'm calling `readInSchema.Equals(inSchema)`. Schema.Equals internally calls Field.Equals. 

I'll push an update to avoid the LINQ calls soon

Reply:
@stephentoub @eerhardt Should there be IEqualityComparer<Schema> / IEqualityComparer<Field> types in order to test schema/field equality? A clear use case would be to check if two schemas are equivalent, although you can make an argument that equivalence and equality are different concerns. A question out of curiosity. Our internal product will eventually involve checking for schema equivalence.

Reply:
I filed https://issues.apache.org/jira/browse/ARROW-4666 to implement IEqualityComparer<Schema/Field>

Reply:
> For why Field needs IEquatable<>, TypeTests::Equality() calls
Assert.True(f0_nullable.Equals(f0_other));

That's a brand new test you're adding, specifically around Equals... so this is saying that the API needs to override Equals because a new test that's being added expects it to override Equals.  That doesn't make sense to me.  Can you clarify?

If your goal is just to validate that two schemas/fields are equal, then you can do that in whatever way the tests desire in those tests, e.g. by creating a method that compares two fields or schemas, or by providing an `IEqualityComparer<Field/Schema>` in the tests.  As @chutchinson suggests, there could even be such a comparer added to the source rather than or in addition to in the tests, but that should be driven by code we actually expect developers to write/need to write against the public APIs, not what a particular test may need to be simpler.  Similarly, it might be desirable for Field/Schema to override Equals/GetHashCode, but that should be driven by what consumers of these APIs need, not by a desire to use them in tests.